text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,321 @@
+  addToHasMany(modelName: string, id?: string, clientId?: string, key: string, jsonApiResources, idx: number) {}
+  removeFromHasMany(modelName: string, id?: string, clientId?: string, key: string, jsonApiResources) {}
+}
+```
+
+The above is just one snapshot of the roughly dozen APIs relating to mutation of the ~25 APIs present on `RecordData`. While these methods
+ serve an important role helping to ensure that all of the features of `ember-data`'s cache, relationship layer, and model layer are available
+ to `RecordData`, the "soup" prevents someone from knowing at a glance which methods are for mutations, and whether these methods are for mutating local state, or communicating remote state changes. It also means that developers must memorize a large API surface area, and don't
+ have a clear picture of how these methods relate to each other in terms of what they achieve.
+
+Contrast that with the following (theoretical) API.
+
+```typescript
+interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
+}
+
+interface Operation {
+  op: string;
+}
+
+interface RecordData {
+  performMutation(operation: Operation): void
+}
+```
+
+We could then describe our original `RecordData` methods above as potential mutation `operation`s.
+For example:
+
+```typescript
+
+interface ReplaceAttributeOperation extends Operation {
+  op: 'replaceAttribute';
+  record: RecordIdentifier;
+  attribute: string;
+  value: any;
+}
+interface AddToRelatedRecordsOperation extends Operation {
+  op: 'addToRelatedRecords';
+  record: RecordIdentifier;
+  relationship: string;
+  relatedRecord: RecordIdentifier;
+}
+interface RemoveFromRelatedRecordsOperation extends Operation {
+  op: 'removeFromRelatedRecords';
+  record: RecordIdentifier;
+  relationship: string;
+  relatedRecord: RecordIdentifier;
+}
+```
+
+The benefit of this structure is that not only does it simplify the cognitive overhead of remembering all of the APIs, it unifies all of our
+ mutations into a single mental model that provides an extension point should we desire to add additional features to `ember-data` such as
+ a log of mutations, reverse operations, or transactions (groups of operations that must succeed or fail together).
+
+### Nested Saves / API Transactions
+
+// TODO 
+
+### Websocket Support
+
+// TODO 
+
+### Better Cache Serialization
+
+// TODO 
+
+### Improved Infra for Offline Support
+
+// TODO 
+
+### Improved Infra for Cache and Relationship Management
+
+// TODO 
+
+----
+
+## Detailed design
+
+### New Typings
+
+```typescript
+import { Value as JSONValue } from 'json-typescript';
+
+export type Meta = { [k: string]: JSONValue };
+
+export interface ResourceIdentifier {
+  lid?: string;
+  id: string | null;
+  type: string;
+  meta?: Meta;
+}
+
+export interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
```

Comment:
similarly i think `meta?` is more appropriate.

Reply:
`?` would imply that the key is sometimes `undefined`. To maintain shape, we always have the key present.